pkg/typesystem/typesystem.go

```diff
@@ -556,18 +556,17 @@ func isUsersetRewriteValid(
 }
 
 func validateRelationTypeRestrictions(model *openfgapb.AuthorizationModel) error {
-	t := New(model)
-	allTupleToUsersetDefinitions := t.GetAllTupleToUsersetsDefinitions()
+	typeSystem := New(model)
 
-	for objectType := range t.typeDefinitions {
-		relations, err := t.GetRelations(objectType)
+	for objectType := range typeSystem.typeDefinitions {
+		relations, err := typeSystem.GetRelations(objectType)
 		if err != nil {
 			return err
 		}
 
 		for name, relation := range relations {
 			relatedTypes := relation.GetTypeInfo().GetDirectlyRelatedUserTypes()
-			assignable := t.IsDirectlyAssignable(relation)
+			assignable := typeSystem.IsDirectlyAssignable(relation)
 
 			if assignable && len(relatedTypes) == 0 {
 				return AssignableRelationError(objectType, name)
```

Comment:
We have very very specific custom errors here that I feel are overkill and add to code bloat for little to no value.

I recommend that we simplify in one of two ways

1) Create a custom error such as `ModelValidationError` and then the error specific message can point out the underlying issue that caused the error such as error messages for un assignable relations or type restrictions that are invalid etc..

2) just return an error with `fmt.Errorf` with an appropriate error message.

We don't consume any specific internal error and switch on or handle the errors differently based on the custom error type, so these custom error types here are just adding to code bloat.

Reply:
Yeah, :100: agree. Let's discuss this tomorrow, then refactor the errors here, in a future PR if necessary.

Reply:
Sounds good 👍